### PR TITLE
Refactor and test response.d

### PR DESCRIPTION
**Pull request by dkorpel**
[Per request](https://forum.dlang.org/thread/qiab48$j46$1@digitalmars.com).

Sorry for the large diff, but the original code was one monster function that did everything. I separated it into three concerns:
- walking the argument list looking for response files to expand
- parsing response files
- looking up response files in the environment / on the filesystem

I added unittests so the first 2 functions are 100% covered. dmd/root/response.d as a whole is 94% covered. Further improvements:

- `core.stdc` imports are now local and selective
- wrong / outdated documentation is updated
- most `goto`s are removed
- use D arrays instead of C pointers
- don't recurse when an `@` in the middle of an argument was found

Note that most behavior remains the same, including:
- response files are allocated and then leak
- FileBuffer is assumed to have a null terminator past the array bounds. (I did add a function `extractDataZ` to at least do this in file.d itself which is less fragile than response.d assuming internals from file.d)
- null terminators are inserted into the response files since arguments are still C-strings.

This is still clumsy, so a possible future PR can add usage of D-strings for arguments as well.

### CODE REVIEW

File: src/dmd/root/file.d [hunk 1]
```diff
@@ -63,6 +63,14 @@ struct File
         {
             return buffer.extractData();
         }
+
+        /// ditto
+        /// Include the null-terminator at the end of the buffer in the returned array.
+        ubyte[] extractDataZ() @nogc nothrow pure
+        {
+            auto result = buffer.extractData();
+            return result.ptr[0 .. result.length + 1];
```
Comment (Geod24) [inline src/dmd/root/file.d]: Interestingly, this isn't covered.
In any case, I think this approach is not the best. Relying on `\0` termination of `char[]` is a bad idea.
I much rather recommend to use `.toCStringThen` when calling a C API function.

File: src/dmd/root/response.d [hunk 2]
```diff
@@ -14,200 +14,341 @@
...
+
 /*********************************
- * #include <stdlib.h>
- * int response_expand(int *pargc,char ***pargv);
- *
  * Expand any response files in command line.
  * Response files are arguments that look like:
  *   @NAME
- * The name is first searched for in the environment. If it is not
- * there, it is searched for as a file name.
```
Comment (Geod24) [inline src/dmd/root/response.d]: I think the lookup order information matters. Can you rephrase it in a way that refers to `lookup` rather than the environment ?

File: src/dmd/root/response.d [hunk 3]
```diff
@@ -14,200 +14,341 @@
...
+ *              arguments in the environment. It is expected to accept a null-
+ *              terminated string and return a mutable char[] that ends with
+ *              a null-terminator or null if the response file could not be
+ *              resolved.
+ *     args = array containing arguments as null-terminated strings
+ *
  * Returns:
- *   0   success
- *   !=0   failure (argc, argv unchanged)
+ *     false on success, true if a response file could not be expanded.
```
Comment (Geod24) [inline src/dmd/root/response.d]: `0` for success makes sense for a C function returning `int`. `false` on success for a D function is heresy (we do commit heresy in DMD, but it's not something we should promote).

File: src/dmd/root/response.d [hunk 4]
```diff
@@ -14,200 +14,341 @@
...
+
+    import core.stdc.stdlib: getenv;
+    import core.stdc.string: strdup, strlen;
+
+    if (auto p = getenv(cp))
     {
-        /* Recursively expand @filename   */
-        if (response_expand(args))
-            goto noexpand;
+        char* buffer = strdup(p);
```
Comment (Geod24) [inline src/dmd/root/response.d]: `Mem.xstrdup`

**Review by Geod24 — CHANGES_REQUESTED** [review]
- Looks like the first commit is wrong / useless. Please squash (as in `git rebase HEAD~2` and use `fixup` to merge the commit together, not the actual `squash`)

File: src/dmd/root/file.d (hunk 1 above)
Comment (dkorpel) [reply src/dmd/root/file.d]: Note that this is the current behavior, it was just hidden in pointer arithmetic before. toCString doesn't work here since the function actually delivers zero-terminated strings by putting null bytes in a mutable buffer. If the argument is at the end of the buffer, this null byte needs to be there.

File: src/dmd/root/response.d (hunk 3 above)
Comment (dkorpel) [reply src/dmd/root/response.d]: I tried to not make any behavioural changes, but I can invert it.

File: src/dmd/root/file.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/root/file.d]: The buffer ends up being stored in `Strings` right ?

Comment (dkorpel) [reply src/dmd/root/file.d]: Yes, and `Strings` == `Array!(const(char)*)`

Comment (Geod24) [reply src/dmd/root/file.d]: I see. I originally thought it was solely to call a C API function, my bad.

File: src/dmd/root/response.d [hunk 5]
```diff
@@ -14,200 +14,344 @@
...
+unittest
+{
+    auto args = Strings(2);
+    args[0] = "@Foo";
+    args[1] = "@Error";
+    assert(!responseExpand!testEnvironment(args));
+}
+
+/*********************************
+ * Take a response-file 'data', parse it and put the resulting
```
Comment (WalterBright) [inline src/dmd/root/response.d]: there are no arguments named 'data'

**Comment by WalterBright** [comment]
Yes, this is a good improvement.

File: src/dmd/root/response.d [hunk 6]
```diff
@@ -14,200 +14,344 @@
 
 module dmd.root.response;
 
-import core.stdc.stdio;
-import core.stdc.stdlib;
-import core.stdc.string;
 import dmd.root.file;
 import dmd.root.filename;
 
+alias responseExpand = responseExpandFrom!lookupInEnvironment;
```
Comment (Geod24) [inline src/dmd/root/response.d]: Add a `///` to make it documented

File: src/dmd/root/response.d [hunk 7]
```diff
@@ -14,200 +14,344 @@
...
+ * Params:
+ *     cp = null-terminated string to look resolve
+ *
+ * Returns:
+ *     a mutable. manually allocated array containing the contents of the environment
+ *     variable or file ending with a null-terminator.
+ *     The null-terminator is inside the bounds of the array.
+ *     If cp could not be resolved, null is returned.
+ */
+char[] lookupInEnvironment(scope const(char)* cp) nothrow {
```
Comment (Geod24) [inline src/dmd/root/response.d]: `private`

File: src/dmd/root/response.d [hunk 8]
```diff
@@ -14,200 +14,344 @@
...
+/*********************************
+ * Try to resolve the null-terminated string cp to a null-terminated char[].
+ * The name is first searched for in the environment. If it is not
+ * there, it is searched for as a file name.
+ *
+ * Params:
+ *     cp = null-terminated string to look resolve
+ *
+ * Returns:
+ *     a mutable. manually allocated array containing the contents of the environment
```
Comment (Geod24) [inline src/dmd/root/response.d]: Is the dot intended ?

File: src/dmd/root/response.d [hunk 9]
```diff
@@ -14,200 +14,344 @@
...
+    assert(index == 2);
+    assert(args[1][0 .. 5] == "ä&#\0");
+
+    testData = "one@\"word \0".dup;
+    assert(insertArgumentsFromResponse(testData, args, index) == false);
+    args[0] = "one@\"word";
+}
+
+/*********************************
+ * Try to resolve the null-terminated string cp to a null-terminated char[].
```
Comment (Geod24) [inline src/dmd/root/response.d]: Needs an empty line after this line to separate title from description

**Review by Geod24 — APPROVED** [review]
LGTM, just a few style nitpick. Ping me when it's fixed and I'll merge.

**Comment by Geod24** [comment]
By the way if you haven't already, I recommend you test it with `dub`, and try to compile, say Vibe.d.
Last time I checked dub was using this feature.

File: src/dmd/root/response.d [hunk 10]
```diff
@@ -14,200 +14,349 @@
...
+    import dmd.root.rmem: mem;
+
+    if (auto p = getenv(cp))
     {
-        /* Recursively expand @filename   */
-        if (response_expand(args))
-            goto noexpand;
+        char* buffer = mem.xstrdup(p);
+        if (!buffer)
+            return null;
```
Comment (Geod24) [inline src/dmd/root/response.d]: This check is redundant: https://github.com/dlang/dmd/blob/b2522da8566783491648bc104a29b42dc2dc569e/src/dmd/root/rmem.d#L39

File: src/dmd/root/file.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/root/file.d]: Hum, was working on something with the same needs, and realized this is wrong.
You don't have the guarantee that `extractData` is `\0` terminated.
See https://github.com/dlang/dmd/pull/10310

Comment (dkorpel) [reply src/dmd/root/file.d]: That PR is about OutBuffer, `extractDataZ` only applies to `ReadResult` of `File` which always allocates even 2 zero-terminators:
https://github.com/dlang/dmd/blob/8bbb2aee973fbd85d3e867285e47883d3ed1c18b/src/dmd/root/file.d#L109
